be/src/storage/segment_sync_executor.cpp

```diff
@@ -0,0 +1,285 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Inc.
+
+#include "storage/segment_sync_executor.h"
+
+#include <fmt/format.h>
+
+#include <memory>
+
+#include "fs/fs_posix.h"
+#include "gen_cpp/data.pb.h"
+#include "runtime/current_thread.h"
+#include "runtime/exec_env.h"
+#include "storage/delta_writer.h"
+#include "util/brpc_stub_cache.h"
+#include "util/raw_container.h"
+
+namespace starrocks {
+
+class SegmentSyncTask final : public Runnable {
+public:
+    SegmentSyncTask(SyncToken* sync_token, std::unique_ptr<SegmentPB> segment, bool eos)
+            : _sync_token(sync_token), _segment(std::move(segment)), _eos(eos) {}
+
+    ~SegmentSyncTask() override = default;
+
+    void run() override { _sync_token->_sync_segment(std::move(_segment), _eos); }
+
+private:
+    SyncToken* _sync_token;
+    std::unique_ptr<SegmentPB> _segment;
+    bool _eos;
+};
+
+SyncChannel::SyncChannel(const DeltaWriterOptions* opt, const std::string& host, int32_t port, int64_t node_id)
+        : _opt(opt), _host(host), _port(port), _node_id(node_id) {
+    _closure = new ReusableClosure<PTabletWriterAddSegmentResult>();
+    _closure->ref();
+}
+
+SyncChannel::~SyncChannel() {
+    if (_closure != nullptr) {
+        _closure->join();
+        if (_closure->unref()) {
+            delete _closure;
+        }
+        _closure = nullptr;
+    }
+}
+
+std::string SyncChannel::debug_string() {
+    return fmt::format("SyncChannnel [host={}, port={}, load_id={}, tablet_id={}, txn_id={}]", _host, _port,
+                       print_id(_opt->load_id), _opt->tablet_id, _opt->txn_id);
+}
+
+Status SyncChannel::_init() {
+    if (_inited) {
+        return Status::OK();
+    }
+    _inited = true;
+
+    _stub = ExecEnv::GetInstance()->brpc_stub_cache()->get_stub(_host, _port);
+    if (_stub == nullptr) {
+        auto msg = fmt::format("Failed to Connect {} failed.", debug_string().c_str());
+        LOG(WARNING) << msg;
+        return Status::InternalError(msg);
+    }
+    return Status::OK();
+}
+
+Status SyncChannel::sync_segment(SegmentPB* segment, butil::IOBuf& data, bool eos,
+                                 std::vector<std::unique_ptr<PTabletInfo>>* sync_tablet_infos) {
+    RETURN_IF_ERROR(_st);
+
+    // 1. init sync channel
+    _st = _init();
+    RETURN_IF_ERROR(_st);
+
+    // 2. send segment sync request
+    _send_request(segment, data, eos);
```

Comment:
Why not capture the returned status?

Reply:
send is async, it will capture in `_wait_response`